gt/communication/ghex_comm_mt.hpp

```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+#include "./ghex_comm.hpp"
+#include "./factorize.hpp"
+#include <ghex/glue/gridtools/field.hpp>
+
+namespace communication {
+
+namespace ghex_comm {
+
+
+class grid_mt
+{
+public: // member types
+  using domain_id_type    = local_domain::domain_id_type;
+  using transport_type    = ::gridtools::ghex::tl::mpi_tag;
+  using grid_type         = ::gridtools::ghex::structured::grid::template type<local_domain>;
+  using patterns_type     = ::gridtools::ghex::pattern_container<transport_type, grid_type, domain_id_type>;
+  using patterns_ptr_t    = std::unique_ptr<patterns_type>;
+
+protected: // member types
+    using coordinate_type = local_domain::coordinate_type;
+    using domain_vec      = std::vector<local_domain>;
+    using comm_obj_type   = typename std::remove_reference<typename std::remove_cv<decltype(
+                               ::gridtools::ghex::make_communication_object<patterns_type>())>::type>::type;
+    using comm_obj_ptr_t  = std::unique_ptr<comm_obj_type>;
+
+public:
+    struct sub_grid
+    {
+        domain_id_type      domain_id;
+        MPI_Comm            mpi_comm;
+        comm_obj_ptr_t      co;
+        patterns_type*      patterns;
+        vec<std::size_t, 2> global_resolution;
+        vec<std::size_t, 2> offset;
+        vec<std::size_t, 3> resolution;
+    };
+
+protected: // members
+    int                 m_size;
+    std::array<int, 2>  m_dims;
+    MPI_Comm            m_comm_cart;
+    int                 m_rank;
+    std::array<int, 2>  m_coords;
+    coordinate_type     m_first;
+    coordinate_type     m_last;
+    halo_generator      m_hg;
+    vec<std::size_t, 2> m_global_resolution;
+    domain_vec          m_domains;
+    patterns_ptr_t      m_patterns;
+    moved_bit           m_moved;
+
+public: // ctors
+    grid_mt(vec<std::size_t, 3> const &global_resolution, int num_sub_domains = 1)
+    : m_size{[this]()
+        {
+            int s;
+            MPI_Comm_size(MPI_COMM_WORLD, &s);
+            return s;
+        }()}
+    , m_dims{[this]()
+        {
+            std::array<int, 2> dims = {0, 0};
+            MPI_Dims_create(this->m_size, 2, dims.data());
+            return dims;
+        }()}
+    , m_comm_cart{[this]()
+        {
+            MPI_Comm comm_;
+            std::array<int, 2> periods = {1, 1};
+            MPI_Cart_create(MPI_COMM_WORLD, 2, m_dims.data(), periods.data(), 1, &comm_);
+            return comm_;
+        }()}
+    , m_rank{[this]()
+        {
+            int r;
+            MPI_Comm_rank(m_comm_cart, &r);
+            return r;
+        }()}
+    , m_coords{[this]()
+        {
+            std::array<int, 2> coords;
+            MPI_Cart_coords(m_comm_cart, m_rank, 2, coords.data());
+            return coords;
+        }()}
+    , m_first{(int)global_resolution.x * m_coords[0] / m_dims[0],
```

Comment:
Is this really what you want (integer division that is truncated)? Don't we get 0 here?

Reply:
improved